app/src/main/kotlin/at/bitfire/davdroid/sync/TaskSyncer.kt

```diff
@@ -11,109 +11,101 @@ import android.content.SyncResult
-                        logger.log(Level.FINE, "Updating local task list $url", info)
-                        list.update(info, updateColors)
-                        // we already have a local task list for this remote collection, don't create a new local task list
-                        newCollections -= url
-                    }
-                }
-
-            // 3. create new local task lists
-            for ((_,info) in newCollections) {
-                logger.log(Level.INFO, "Adding local task list", info)
-                LocalTaskList.create(account, taskProvider, info)
-            }
-
-            // 4. sync local task lists
-            val localTaskLists = DmfsTaskList
-                .find(account, taskProvider, LocalTaskList.Factory, "${TaskContract.TaskLists.SYNC_ENABLED}!=0", null)
-            for (localTaskList in localTaskLists) {
-                logger.info("Synchronizing task list #${localTaskList.id} [${localTaskList.syncId}]")
-
-                val url = localTaskList.syncId?.toHttpUrl()
-                remoteCollections[url]?.let { collection ->
-                    val syncManager = tasksSyncManagerFactory.tasksSyncManager(
-                        account,
-                        accountSettings,
-                        httpClient.value,
-                        extras,
-                        authority,
-                        syncResult,
-                        localTaskList,
-                        collection
-                    )
-                    syncManager.performSync()
-                }
-            }
+): Syncer<LocalTaskList>(account, extras, syncResult) {
+
+    @AssistedFactory
+    interface Factory {
+        fun create(account: Account, authority: String, extras: Array<String>, syncResult: SyncResult): TaskSyncer
+    }
+
+    private lateinit var taskProvider: TaskProvider
+
+    override val serviceType: String
+        get() = Service.TYPE_CALDAV
+    override fun localCollections(provider: ContentProviderClient): List<LocalTaskList>
+        = DmfsTaskList.find(account, taskProvider, LocalTaskList.Factory, null, null)
+    override fun localSyncCollections(provider: ContentProviderClient): List<LocalTaskList>
+        = DmfsTaskList.find(account, taskProvider, LocalTaskList.Factory, "${TaskContract.TaskLists.SYNC_ENABLED}!=0", null)
+
+    override fun beforeSync(provider: ContentProviderClient) {
+
+        // Acquire task provider
+        val providerName = TaskProvider.ProviderName.fromAuthority(authority)
+        taskProvider = try {
+            TaskProvider.fromProviderClient(context, providerName, provider)
         } catch (e: TaskProvider.ProviderTooOldException) {
             tasksAppManager.get().notifyProviderTooOld(e)
             syncResult.databaseError = true
-        } catch (e: Exception) {
-            logger.log(Level.SEVERE, "Couldn't sync task lists", e)
-            syncResult.databaseError = true
+            return // Don't sync
```

Comment:
Does this "don't sync" do anything? Especially does it prevent the uninitialized `taskProvider` from being accessed?

Reply:
It should prevent the whole sync from running. I added a boolean return.